onnxruntime/core/optimizer/skip_layer_norm.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/skip_layer_norm_fusion.h"
+#include "core/graph/graph_utils.h"
+#include "float.h"
+#include <deque>
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+namespace onnxruntime {
+
+// LayerNorm supports limited data types.
+static std::vector<std::string> supported_data_types{"tensor(float16)", "tensor(float)", "tensor(double)"};
+
+static bool IsSupportedDataType(const Node& node) {
+  for (const auto& input_arg : node.InputDefs()) {
+    if (std::find(supported_data_types.begin(), supported_data_types.end(),
+                  *(input_arg->Type())) == supported_data_types.end()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+Skip Layer Normalization will fuse Add + LayerNormalization into one node
+*/
+Status SkipLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  std::vector<std::reference_wrapper<Node>> nodes_to_remove;
+  for (auto node_index : node_topology_list) {
+    nodes_to_remove.clear();
+    auto* p_add = graph.GetNode(node_index);
+    if (p_add == nullptr)
+      continue;  // we removed the node as part of an earlier fusion
+
+    Node& add_node = *p_add;
+    ORT_RETURN_IF_ERROR(Recurse(add_node, modified, graph_level));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(add_node, "Add", {7}) ||
+        !graph_utils::IsSupportedProvider(add_node, GetCompatibleExecutionProviders()) ||
+        add_node.GetOutputEdgesCount() != 1 ||
+        !IsSupportedDataType(add_node)) {
+      continue;
+    }
+    nodes_to_remove.push_back(add_node);
+
+    // Find "LayerNormalization" node after the "Add"
+    Node& ln_node = *graph.GetNode(add_node.OutputNodesBegin()->Index());
```

Comment:
Please also check the dimension of Add. Input 0 and 1 shall be same shape, and 3D.
    // Input and output shapes
    //   Input 0 - input: (batch_size, sequence_length, hidden_size)
    //   Input 1 - skip : (batch_size, sequence_length, hidden_size)
    //   Input 2 - gamma: (hidden_size)
    //   Input 3 - beta : (hidden_size)